docs/cryptodoc/src/05_10_cmce.rst

```diff
@@ -0,0 +1,485 @@
+key decapsulation. One of the crucial steps in the decapsulation algorithm of
+Classic McEliece is the decoding subroutine described in Section 7.4 of
+[CMCE-ISO]_. This subroutine is implemented based on the recommendations
+provided in Section 6 of [McBits]_. It utilizes Berlekamp's
+algorithm for Goppa decoding.
+
+To begin with, the code word :math:`C` that needs to be decoded is extended by
+appending zeros. This results in a binary vector :math:`v = (C,0,\dots,0) \in \mathbb{F}_2^{n}`,
+as Step 1 of Section 7.2 of [CMCE-ISO]_ defines. Subsequently, the syndrome for
+Berlekamp's method is computed from :math:`v`. The syndrome is a vector given by
+:math:`\left( \sum\nolimits_{i} \frac{v_i\alpha_i^0}{g(\alpha_i)^2},\dots,\sum\nolimits_{i} \frac{v_i\alpha_i^{n-1}}{g(\alpha_i)^2} \right)`,
+where :math:`\alpha_i` are the first :math:`n` field ordering elements and
+:math:`g` is the Goppa polynomial.
+
+Next, an error locator polynomial :math:`\sigma` is computed using the
+Berlekamp-Massey algorithm on the syndrome. The resulting polynomial has a
+particular property that allows deriving the error vector :math:`e`.
+Specifically, :math:`\sigma(\alpha_i) = 0` if and only if
+:math:`e_i = 1`. By evaluating :math:`\sigma` at :math:`\alpha_0,\dots,\alpha_{n-1}`,
+we can reconstruct the error vector :math:`e`.
+
+To ensure accurate decoding, Botan follows the recommendation of [CMCE-IMPL]_ Section 6.3.
+It computes the syndrome for the error
+vector :math:`e` and compares it with the syndrome for :math:`v`. If the
+comparison holds and the weight of :math:`e` is equal to :math:`t`, we consider
+the decoding successful. Otherwise, it is flagged as a failure.
+
+It is worth noting that the syndrome computation, Berlekamp-Massey algorithm, and
+locator polynomial evaluation are implemented in constant time. Additionally,
+the checks for the weight of :math:`e` and the syndrome comparison are designed
+to avoid early abortion if any check fails. This ensures that no information about
+the decoding success leaks.
+
+
+.. _pubkey/cmce/params:
+
+Parameters
+^^^^^^^^^^
+
+The ``Classic_McEliece_Parameter_Set`` enum contains all instances of the
+Classic McEliece algorithm listed in Table :ref:`Supported Classic McEliece
+Parameter Sets <pubkey/cmce/parameter_table>`. These parameter sets serve as
+the basis for deriving all the necessary parameters used within the algorithm.
+They are collected in a ``Classic_McElice_Parameters`` object, which includes
+all the parameters defined in the specifications [CMCE-ISO]_ and [CMCE-R4]_.
+This object is passed to all algorithm components, ensuring consistent
+parameter usage.
+
+
+.. _pubkey/cmce/key_gen:
+
+
+Key Generation
+--------------
+
+Classic McEliece key generation follows Section 8.3 of [CMCE-ISO]_ and is
+implemented within the ``Classic_McEliece_KeyPair_Internal`` class. It works as
+follows:
+
+.. admonition:: Classic McEliece Key Generation
+
+   **Input:**
```

Comment:
From our side, it would be useful to list parameters or mode specifications as an input in the documentation when they are also required as an input to the respective implemented function like it is for Classic McEliece.